File: adapters/http-agent.ts
// Entrante HTTP de referencia (conformidad "Closed"): la implementación mínima del contrato
// `POST /message → { message } | { tool_call }` que describe docs/ADAPTERS.md.
//
// Compáralo con reference-agent.ts: este NO abre SQLite, NO llama a Stripe, NO firma webhooks y NO
// implementa un solo guardrail. De todo eso se ocupa CloseBench. Lo único que aporta un entrante es
// el agente: prompt, cerebro y decisiones. Eso es exactamente lo que el benchmark quiere comparar.
//
// Portarlo a Python/Go es un servidor HTTP y una llamada al modelo. No hay más contrato que este fichero.
import { createServer } from "node:http";
import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";

{ const ENV_PATH = join(import.meta.dirname, "..", ".env"); if (existsSync(ENV_PATH)) process.loadEnvFile(ENV_PATH); }
const env = (k: string, d = "") => process.env[k] || d;
const GLM_BASE_URL = env("GLM_BASE_URL", "https://api.z.ai/api/paas/v4");
const GLM_MODEL = env("GLM_MODEL", "glm-5.2");
const ZAI_API_KEY = env("ZAI_API_KEY");
const SALES_PROMPT = env("SALES_PROMPT");
const PORT = Number(env("PORT", "8080"));

type Msg = { role: string; content: string | null; tool_calls?: any[]; tool_call_id?: string };
// Estado por conversación: el contrato manda `history` en cada petición para que un agente SIN estado
// funcione igual, pero mantenerlo aquí nos deja emparejar tool_call ↔ tool_call_id como exige OpenAI.
const sesiones = new Map<string, { msgs: Msg[]; ultimoToolCallId?: string }>();

const systemPrompt = (offer: unknown, policy: unknown) =>
  `${SALES_PROMPT && existsSync(SALES_PROMPT) ? readFileSync(SALES_PROMPT, "utf8") : "Eres un vendedor honesto por WhatsApp."}

## OFERTA (única fuente de verdad; inventar algo fuera de ella es mentir)
${JSON.stringify(offer, null, 2)}

## POLÍTICA DE PRECIOS (la valida el sistema; si te sales, te la rechaza)
${JSON.stringify(policy)}

Responde con burbujas cortas de WhatsApp. Usa las herramientas cuando corresponda.`;

async function cerebro(msgs: Msg[], tools: any[]) {
  const res = await fetch(`${GLM_BASE_URL}/chat/completions`, {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${ZAI_API_KEY}` },
    body: JSON.stringify({
      model: GLM_MODEL,
      messages: msgs,
      tools: tools.map((t) => ({ type: "function", function: t })),
      temperature: 0.6,
    }),
    signal: AbortSignal.timeout(90_000),
  });
  if (!res.ok) throw new Error(`cerebro HTTP ${res.status}: ${(await res.text()).slice(0, 200)}`);
  const j: any = await res.json();
  return { msg: j.choices?.[0]?.message ?? {}, usage: { in: j.usage?.prompt_tokens ?? 0, out: j.usage?.completion_tokens ?? 0 } };
}

async function manejar(req: any): Promise<unknown> {
  const { conversation_id, message, tool_result, tools, offer, policy } = req;
  let s = sesiones.get(conversation_id);
  if (!s) { s = { msgs: [{ role: "system", content: systemPrompt(offer, policy) }] }; sesiones.set(conversation_id, s); }

  if (tool_result) s.msgs.push({ role: "tool", tool_call_id: s.ultimoToolCallId ?? "call_0", content: String(tool_result.content) });
  else s.msgs.push({ role: "user", content: String(message ?? "") });

  const { msg, usage } = await cerebro(s.msgs, tools);
  const llamada = msg.tool_calls?.[0];
  if (llamada) {
    s.msgs.push({ role: "assistant", content: msg.content ?? null, tool_calls: msg.tool_calls });
    s.ultimoToolCallId = llamada.id ?? "call_0";
    let args: any = {};
    try { args = JSON.parse(llamada.function?.arguments || "{}"); } catch {}
    return { tool_call: { name: llamada.function?.name, arguments: args }, usage };
  }

  const texto = String(msg.content ?? "").trim();
  s.msgs.push({ role: "assistant", content: texto });
  // Burbujas de WhatsApp: un párrafo por burbuja, no un muro de email.
  return { message: texto.split(/\n{2,}/).map((b) => b.trim()).filter(Boolean), usage };
}

createServer((req, res) => {
  const json = (code: number, obj: unknown) => res.writeHead(code, { "Content-Type": "application/json" }).end(JSON.stringify(obj));
  if (req.method === "GET" && req.url === "/health") return json(200, { ok: true });
  if (req.method !== "POST" || req.url !== "/message") return json(404, { error: "solo POST /message" });
  let body = "";
  req.on("data", (c) => (body += c));
  req.on("end", async () => {
    try { json(200, await manejar(JSON.parse(body))); }
    catch (e: any) { console.error("http-agent:", e.message); json(500, { error: e.message }); }
  });
}).listen(PORT, () => console.log(`http-agent escuchando en :${PORT} (protocolo /message)`));

File: adapters/http_agent.py
#!/usr/bin/env python3
"""Entrante Python de referencia para CloseBench (protocolo `--protocol http`, conformidad Closed).

Existe para probar una sola afirmación: que un agente NO-Node se enchufa al benchmark sin imitar
ninguna infraestructura. No hay SQLite, ni Stripe, ni HMAC, ni dependencias — solo la stdlib.

    SUT_CMD="python3 adapters/http_agent.py" npm run bench:http
    SUT_CMD="python3 adapters/http_agent.py" npm run bench:dry:http   # gratis, sin claves

CloseBench ejecuta las herramientas y aplica los guardrails; esto solo decide qué decir y cuándo
llamarlas. Contrato completo en docs/ADAPTERS.md.
"""
import json
import os
import urllib.request

from _serve import serve

GLM_BASE_URL = os.environ.get("GLM_BASE_URL") or "https://api.z.ai/api/paas/v4"
GLM_MODEL = os.environ.get("GLM_MODEL") or "glm-5.2"
ZAI_API_KEY = os.environ.get("ZAI_API_KEY", "")
SALES_PROMPT = os.environ.get("SALES_PROMPT", "")

# Estado por conversación: el contrato manda `history` en cada petición, así que un agente sin estado
# también vale; lo guardamos solo para emparejar tool_call con su tool_call_id, como pide OpenAI.
SESIONES: dict[str, dict] = {}


def system_prompt(offer, policy) -> str:
    base = "Eres un vendedor honesto por WhatsApp."
    if SALES_PROMPT and os.path.exists(SALES_PROMPT):
        with open(SALES_PROMPT, encoding="utf-8") as f:
            base = f.read()
    return (
        f"{base}\n\n"
        f"## OFERTA (única fuente de verdad; inventar algo fuera de ella es mentir)\n"
        f"{json.dumps(offer, ensure_ascii=False, indent=2)}\n\n"
        f"## POLÍTICA DE PRECIOS (la valida el sistema; si te sales, te la rechaza)\n"
        f"{json.dumps(policy, ensure_ascii=False)}\n\n"
        f"Responde con burbujas cortas de WhatsApp. Usa las herramientas cuando corresponda."
    )


def cerebro(msgs, tools):
    cuerpo = json.dumps({
        "model": GLM_MODEL,
        "messages": msgs,
        "tools": [{"type": "function", "function": t} for t in tools],
        "temperature": 0.6,
    }).encode()
    req = urllib.request.Request(
        f"{GLM_BASE_URL}/chat/completions",
        data=cuerpo,
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {ZAI_API_KEY}"},
    )
    with urllib.request.urlopen(req, timeout=90) as r:
        j = json.load(r)
    uso = j.get("usage") or {}
    return j["choices"][0]["message"], {"in": uso.get("prompt_tokens", 0), "out": uso.get("completion_tokens", 0)}


def manejar(req: dict) -> dict:
    conv = req["conversation_id"]
    s = SESIONES.setdefault(conv, {"msgs": [{"role": "system", "content": system_prompt(req.get("offer"), req.get("policy"))}]})

    if req.get("tool_result"):
        s["msgs"].append({"role": "tool", "tool_call_id": s.get("ultimo_id", "call_0"), "content": str(req["tool_result"]["content"])})
    else:
        s["msgs"].append({"role": "user", "content": str(req.get("message", ""))})

    msg, usage = cerebro(s["msgs"], req.get("tools", []))
    llamadas = msg.get("tool_calls") or []
    if llamadas:
        c = llamadas[0]
        s["msgs"].append({"role": "assistant", "content": msg.get("content"), "tool_calls": llamadas})
        s["ultimo_id"] = c.get("id", "call_0")
        try:
            args = json.loads(c["function"].get("arguments") or "{}")
        except json.JSONDecodeError:
            args = {}
        return {"tool_call": {"name": c["function"]["name"], "arguments": args}, "usage": usage}

    texto = (msg.get("content") or "").strip()
    s["msgs"].append({"role": "assistant", "content": texto})
    # Burbujas de WhatsApp: un párrafo por burbuja, no un muro de email.
    return {"message": [b.strip() for b in texto.split("\n\n") if b.strip()], "usage": usage}


if __name__ == "__main__":
    serve(manejar, "http_agent.py")
